Plugins/BuildServerIntegration/JenkinsIntegration/JenkinsAdapter.cs

```diff
@@ -114,27 +123,25 @@ public IObservable<BuildInfo> GetRunningBuilds(IScheduler scheduler)
 
         public IObservable<BuildInfo> GetBuilds(IScheduler scheduler, DateTime? sinceDate = null, bool? running = null)
         {
-            if (_getBuildUrls == null || _getBuildUrls.Count() == 0)
-            {
-                return Observable.Empty<BuildInfo>(scheduler);
-            }
-
             return Observable.Create<BuildInfo>((observer, cancellationToken) =>
                 Task<IDisposable>.Factory.StartNew(
                     () => scheduler.Schedule(() => ObserveBuilds(sinceDate, running, observer, cancellationToken))));
         }
 
+        private static Dictionary<string, BuildInfo> _finishedBuildsInfo = new Dictionary<string, BuildInfo>();
+
         private void ObserveBuilds(DateTime? sinceDate, bool? running, IObserver<BuildInfo> observer, CancellationToken cancellationToken)
         {
             try
             {
-                if (_getBuildUrls.All(t => t.IsCanceled))
+                var buildUrls = getBuildUrls(running ?? false);
```

Comment:
lines 138 and 144 cause multiple enumerations. 
perhaps change the signature of `GetBuildUrls` to be `IList<T>` or `T[]`